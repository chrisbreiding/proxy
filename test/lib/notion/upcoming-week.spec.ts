import nock from 'nock'
import { afterEach, describe, expect, it } from 'vitest'

import { handleServer } from '../../support/setup'
import {
  nockAppendBlockChildren,
  nockGetBlockChildren,
  nockUpdateBlock,
  snapshotBody,
} from '../../support/util'
import { startServer } from '../../../index'

process.env.API_KEY = 'key'

describe('lib/notion/upcoming-week', () => {

  handleServer(startServer)

  describe('GET /notion/upcoming-week/:key', () => {
    it('renders upcoming week button', async (ctx) => {
      const res = await ctx.request.get('/notion/upcoming-week/key')

      expect(res.status).to.equal(200)
      expect(res.headers['content-type']).to.equal('text/html; charset=utf-8')
      expect(res.text).to.include('<button class="add">Add Following Week</button>')
    })

    it('status 403 if key does not match', async (ctx) => {
      const res = await ctx.request.get('/notion/upcoming-week/nope')

      expect(res.status).to.equal(403)
    })
  })

  describe('POST /notion/upcoming-week/:key', () => {
    afterEach(() => {
      nock.cleanAll()
    })

    it('adds week based on template and updates button url', async (ctx) => {
      nockGetBlockChildren('week-template-id', { fixture: 'upcoming-week/week-template-blocks' })
      nockGetBlockChildren('nested-parent-id', { fixture: 'blocks' })

      const snapshots = [
        snapshotBody(nockAppendBlockChildren({
          id: 'append-to-id',
          fixture: 'upcoming-week/append-1-result',
        }), 'append-1'),
        snapshotBody(nockAppendBlockChildren({ id: 'nested-parent-id' }), 'nested'),
        snapshotBody(nockAppendBlockChildren({ id: 'append-to-id' }), 'append-2'),
        snapshotBody(nockUpdateBlock('button-id'), 'button'),
      ]

      const query = [
        ['weekTemplatePageId', 'week-template-id'],
        ['appendToId', 'append-to-id'],
        ['addFollowingWeekButtonId', 'button-id'],
        ['notionToken', 'notion-token'],
        ['startDate', '2022-11-06T12:00:00.000Z'],
      ].map(([key, value]) => `${key}=${value}`).join('&')

      const res = await ctx.request.post(`/notion/upcoming-week/key?${query}`)

      expect(res.status).to.equal(200)
      expect(res.headers['content-type']).to.equal('text/html; charset=utf-8')
      expect(res.text).to.include('Following week successfully added!')

      await Promise.all(snapshots)
    })

    it('status 403 if key does not match', async (ctx) => {
      const res = await ctx.request.post('/notion/upcoming-week/nope')

      expect(res.status).to.equal(403)
    })
  })
})
