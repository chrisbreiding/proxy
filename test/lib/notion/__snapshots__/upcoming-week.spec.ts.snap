// Vitest Snapshot v1

exports[`lib/notion/upcoming-week > POST /notion/upcoming-week/:key > adds week based on template and updates button url > append-1 1`] = `
{
  "after": "last-upcoming-id",
  "children": [
    {
      "object": "block",
      "paragraph": {
        "rich_text": [
          {
            "plain_text": "Sun, 12/26",
            "text": {
              "content": "Sun, 12/26",
            },
            "type": "text",
          },
        ],
      },
      "type": "paragraph",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Value 1",
            "text": {
              "content": "Value 1",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Sun task 2",
            "text": {
              "content": "Sun task 2",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Sun conditional task that matches",
            "text": {
              "content": "Sun conditional task that matches",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "object": "block",
      "paragraph": {
        "rich_text": [
          {
            "plain_text": "Mon, 12/27",
            "text": {
              "content": "Mon, 12/27",
            },
            "type": "text",
          },
        ],
      },
      "type": "paragraph",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Mon Value 2: with a colon",
            "text": {
              "content": "Mon Value 2: with a colon",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Mon task 2",
            "text": {
              "content": "Mon task 2",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "object": "block",
      "paragraph": {
        "rich_text": [
          {
            "plain_text": "Tue, 12/28",
            "text": {
              "content": "Tue, 12/28",
            },
            "type": "text",
          },
        ],
      },
      "type": "paragraph",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Tue task 1",
            "text": {
              "content": "Tue task 1",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "object": "block",
      "paragraph": {
        "rich_text": [
          {
            "plain_text": "Wed, 12/29",
            "text": {
              "content": "Wed, 12/29",
            },
            "type": "text",
          },
        ],
      },
      "type": "paragraph",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Wed task 1 (has children)",
            "text": {
              "content": "Wed task 1 (has children)",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
  ],
}
`;

exports[`lib/notion/upcoming-week > POST /notion/upcoming-week/:key > adds week based on template and updates button url > append-2 1`] = `
{
  "after": "nested-parent-id",
  "children": [
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Wed task 2",
            "text": {
              "content": "Wed task 2",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "object": "block",
      "paragraph": {
        "rich_text": [
          {
            "plain_text": "Thu, 12/30",
            "text": {
              "content": "Thu, 12/30",
            },
            "type": "text",
          },
        ],
      },
      "type": "paragraph",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Thu task 1",
            "text": {
              "content": "Thu task 1",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "object": "block",
      "paragraph": {
        "rich_text": [
          {
            "plain_text": "Fri, 12/31",
            "text": {
              "content": "Fri, 12/31",
            },
            "type": "text",
          },
        ],
      },
      "type": "paragraph",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Value 3 (suffix)",
            "text": {
              "content": "Value 3 (suffix)",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "object": "block",
      "paragraph": {
        "rich_text": [
          {
            "plain_text": "Sat, 1/1",
            "text": {
              "content": "Sat, 1/1",
            },
            "type": "text",
          },
        ],
      },
      "type": "paragraph",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Sat task 1",
            "text": {
              "content": "Sat task 1",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Sat task 2",
            "text": {
              "content": "Sat task 2",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
  ],
}
`;

exports[`lib/notion/upcoming-week > POST /notion/upcoming-week/:key > adds week based on template and updates button url > button 1`] = `
{
  "embed": {
    "caption": [],
    "url": "https://proxy.crbapps.com/notion/upcoming-week/key?addFollowingWeekButtonId=button-id&notionToken=notion-token&startDate=2022-01-02T12:00:00.000Z&upcomingId=upcoming-id&weekTemplatePageId=week-template-id",
  },
  "object": "block",
  "type": "embed",
}
`;

exports[`lib/notion/upcoming-week > POST /notion/upcoming-week/:key > adds week based on template and updates button url > nested 1`] = `
{
  "children": [
    {
      "bulleted_list_item": {
        "color": "default",
        "rich_text": [
          {
            "annotations": {
              "bold": false,
              "code": false,
              "color": "default",
              "italic": false,
              "strikethrough": false,
              "underline": false,
            },
            "href": null,
            "plain_text": "Task 1",
            "text": {
              "content": "Task 1",
              "link": null,
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "bulleted_list_item": {
        "color": "default",
        "rich_text": [
          {
            "annotations": {
              "bold": false,
              "code": false,
              "color": "default",
              "italic": false,
              "strikethrough": false,
              "underline": false,
            },
            "href": null,
            "plain_text": "Task 2",
            "text": {
              "content": "Task 2",
              "link": null,
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
  ],
}
`;

exports[`lib/notion/upcoming-week > POST /notion/upcoming-week/:key > gracefully handles absence of button block and variables block > append-1 1`] = `
{
  "children": [
    {
      "object": "block",
      "paragraph": {
        "rich_text": [
          {
            "plain_text": "Sun, 12/26",
            "text": {
              "content": "Sun, 12/26",
            },
            "type": "text",
          },
        ],
      },
      "type": "paragraph",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "\${var1}",
            "text": {
              "content": "\${var1}",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Sun task 2",
            "text": {
              "content": "Sun task 2",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Sun conditional task that matches",
            "text": {
              "content": "Sun conditional task that matches",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
    {
      "object": "block",
      "paragraph": {
        "rich_text": [
          {
            "plain_text": "Mon, 12/27",
            "text": {
              "content": "Mon, 12/27",
            },
            "type": "text",
          },
        ],
      },
      "type": "paragraph",
    },
    {
      "bulleted_list_item": {
        "rich_text": [
          {
            "plain_text": "Mon \${var2}",
            "text": {
              "content": "Mon \${var2}",
            },
            "type": "text",
          },
        ],
      },
      "object": "block",
      "type": "bulleted_list_item",
    },
  ],
}
`;
